### PR TITLE
eMRTD additions

### PR DESCRIPTION
**Pull request by aveao**
This PR:

- Dumps EF_CardAccess when doing hf emrtd dump
- Automatically changes document number to uppercase and checks for length of DOB/Expiry to ensure that stuff that has no chance of working will not be sent to the eMRTD

More information at #1117.

### CODE REVIEW

File: client/src/cmdhfemrtd.c [hunk 1]
```diff
@@ -1296,6 +1291,15 @@ int infoHF_EMRTD(char *documentnumber, char *dob, char *expiry, bool BAC_availab
...
 
+static void text_to_upper(uint8_t *data, int datalen) {
+    // Loop over text to make lowercase text uppercase
+    for (int i = 0; i < datalen; i++) {
+        if ('a' <= data[i] && data[i] <= 'z') {
+            data[i] -= 32;
+        }
+    }
+}
+
```
Comment (iceman1001) [inline client/src/cmdhfemrtd.c:1302]: there is a  strup function to use...

Comment (aveao) [reply client/src/cmdhfemrtd.c:1302]: I checked the code and couldn't find something like that, where is it?

File: client/src/cmdhfemrtd.c [hunk 2]
```diff
@@ -1329,10 +1334,22 @@ static int cmd_hf_emrtd_dump(const char *Cmd) {
     
     if (CLIParamStrToBuf(arg_get_str(ctx, 2), dob, 6, &slen) != 0 || slen == 0) {
         BAC = false;
-    } 
+    } else {
+        if (slen != 6) {
```
Comment (iceman1001) [inline client/src/cmdhfemrtd.c:1338]: maybe not only check len,  but also if  MM is <=12 and DD <=31  (generic naive test)
or have  a array with MM, maxdd and make a lookup..

Comment (aveao) [reply client/src/cmdhfemrtd.c:1338]: Will do.

File: client/src/cmdhfemrtd.c (hunk 1 above)
Comment (doegox) [reply client/src/cmdhfemrtd.c:1302]: toupper

**Comment by aveao** [comment]
Resolved both issues. Input welcome on how I could handle that date thing better fwiw, strtol might not be ideal.

**Comment by aveao** [comment]
The EF_CardAccess dump isn't working, will sort that out and then unmark as draft.

**Comment by aveao** [comment]
Fixed, should be ready to re-review.

I've also added in some code improvements recommended by @ceres-c, thank you so much for all the help.

File: client/src/cmdhfemrtd.c [hunk 3]
```diff
@@ -711,18 +710,17 @@ static bool emrtd_select_and_read(uint8_t *dataout, int *dataoutlen, const char
...
-    for (int i = 0; i < file_length - 6; i++) {
-        if ((file_contents[i] == 0xFF && file_contents[i + 1] == 0xD8 && file_contents[i + 2] == 0xFF && file_contents[i + 3] == 0xE0) ||
-            (file_contents[i] == 0x00 && file_contents[i + 1] == 0x00 && file_contents[i + 2] == 0x00 && file_contents[i + 3] == 0x0C && file_contents[i + 4] == 0x6A && file_contents[i + 5] == 0x50)) {
-            datalen = file_length - i;
-            memcpy(data, file_contents + i, datalen);
+    // Note: Doing file_length - 6 to account for the longest data we're checking.
+    for (offset = 0; offset < file_length - 6; offset++) {
+        if ((file_contents[offset] == 0xFF && file_contents[offset + 1] == 0xD8 && file_contents[offset + 2] == 0xFF && file_contents[offset + 3] == 0xE0) ||
+            (file_contents[offset] == 0x00 && file_contents[offset + 1] == 0x00 && file_contents[offset + 2] == 0x00 && file_contents[offset + 3] == 0x0C && file_contents[offset + 4] == 0x6A && file_contents[offset + 5] == 0x50)) {
+            datalen = file_length - offset;
```
Comment (iceman1001) [inline client/src/cmdhfemrtd.c:723]: Use memcmp  fct ?

Comment (aveao) [reply client/src/cmdhfemrtd.c:723]: Considered that, would likely have a slightly higher overhead, but I'll still do it for better visibility later today.

**Pull request merged**